CorsixTH/Lua/dialogs/edit_room.lua

```diff
@@ -1086,71 +1114,90 @@ local function checkDoorWalls(x, y, wall, has_swingdoor)
     dx = 1
     dy = 0
   end
-
+  local invalid_tile = 0
   if th:getCell(x, y, wall_num) % 0x100 ~= 0 then
-    return false
+    invalid_tile = 4
   end
 
   -- If it is a swing door there are two more locations to check.
   if has_swingdoor then
-    if th:getCell(x - dx, y - dy, wall_num) % 0x100 ~= 0 or
-        th:getCell(x + dx, y + dy, wall_num) % 0x100 ~= 0 then
-      return false
+    if th:getCell(x - dx, y - dy, wall_num) % 0x100 ~= 0 then
+      invalid_tile = invalid_tile + 2
+    end
+    if th:getCell(x + dx, y + dy, wall_num) % 0x100 ~= 0 then
+      invalid_tile = invalid_tile + 8
     end
   end
-  return true
+  return invalid_tile
 end
 
 --! Check whether the given tile can function as a door entry/exit tile.
 --!param xpos (int) X position of the tile.
 --!param ypos (int) Y position of the tile.
 --!param player_id (int) Player id owning the hospital.
---!param flag_names (array) If set, array with two additional required properties.
+--!param world - reference to world object instance
 --!return Whether the tile is considered to be valid.
-local function validDoorTile(xpos, ypos, player_id, flag_names)
+local function validDoorTile(xpos, ypos, player_id, world)
   local th = TheApp.map.th
-
   local tile_flags = th:getCellFlags(xpos, ypos)
-  if not (tile_flags.buildable or tile_flags.passable or tile_flags.owner == player_id) then return false end
-  if not flag_names then return true end
-  return tile_flags[flag_names[1]] and tile_flags[flag_names[2]]
+  -- check builable and own it
+  if not (tile_flags.buildable or tile_flags.owner == player_id) then return false end
+  -- any object will cause it to be blocked (ignore litter)
+  if not(tile_flags.thob == 0 or tile_flags.thob == 62) then return false end
+  -- check if its passable that no object footprint blocks it
+  if tile_flags.passable then return world:isTileExclusivelyPassable(xpos, ypos, 1) end
+  return true
 end
 
 function UIEditRoom:setDoorBlueprint(orig_x, orig_y, orig_wall)
-  local x = orig_x
-  local y = orig_y
-  local wall = orig_wall
-
-  -- Used to get the adjacent tiles when placing swing doors.
-  local x_mod
-  local y_mod
-  if wall == "south" then
-    y = y + 1
-    wall = "north"
-    x_mod = 2
-  elseif wall == "east" then
-    x = x + 1
-    wall = "west"
-    y_mod = 2
-  elseif wall == "north" then
-    x_mod = 2
-  else
-    y_mod = 2
+  --! Calculate position offsets and door blueprint wall values
+  --! param x - doors blueprint x value
+  --! param y - doors blueprint y value
+  --! param wall - original wall orientation
+  --! return x - updated x value
+  --! return y - updated y value
+  --! return x_mod - offest value to apply to tile count to determine relative position
+  --! return y_mod - offset value to apply to tile count to determine relatitve position
+  --! return wall - wall orientation style (only 2 styles)
+  local function doorWallOffsetCalculations(x, y, wall)
```

Comment:
Move this function up to file scope (out of the setDoorBlueprint function). We are progressing to a style where we avoid nested functions unless closure is needed.